File: agenkit-scala/src/main/scala/io/agenkit/adapters/OpenAiAdapter.scala
package io.agenkit.adapters

import io.agenkit.core.Message
import scala.concurrent.{ExecutionContext, Future}

/** OpenAI Chat Completions adapter.
  *
  * Calls `POST {baseUrl}/chat/completions` with bearer auth and maps the first
  * choice's message back into an agenkit [[Message]]. Works against OpenAI and
  * any OpenAI-compatible endpoint (vLLM, llama.cpp, SGLang, …) via `baseUrl`.
  */
class OpenAiAdapter(
  apiKey: String,
  model: String = "gpt-4o",
  baseUrl: String = "https://api.openai.com/v1"
) extends LlmClient:

  def complete(messages: List[Message])(using ExecutionContext): Future[Message] =
    val payload = ujson.Obj(
      "model" -> model,
      "messages" -> ujson.Arr.from(
        messages.map(m => ujson.Obj("role" -> m.role, "content" -> m.contentString))
      )
    )

    HttpLlm
      .postJson(
        s"$baseUrl/chat/completions",
        ujson.write(payload),
        Map("authorization" -> s"Bearer $apiKey")
      )
      .map { body =>
        val json = ujson.read(body)
        val choices = json("choices").arr
        if choices.isEmpty then throw new RuntimeException("OpenAI response contained no choices")
        val content = choices(0)("message")("content").str
        Message.of("assistant", content)
      }

File: agenkit-java/src/main/java/io/agenkit/adapters/OpenAiAdapter.java
package io.agenkit.adapters;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.agenkit.core.Message;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * LLM client for OpenAI-compatible APIs.
 */
public final class OpenAiAdapter implements LlmClient {

    private static final Logger log = LoggerFactory.getLogger(OpenAiAdapter.class);

    private final String apiKey;
    private final String model;
    private final String baseUrl;
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;

    public OpenAiAdapter(String apiKey, String model) {
        this(apiKey, model, "https://api.openai.com/v1");
    }

    public OpenAiAdapter(String apiKey, String model, String baseUrl) {
        this.apiKey = apiKey;
        this.model = model;
        this.baseUrl = baseUrl;
        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(30))
                .build();
        this.objectMapper = new ObjectMapper();
    }

    @Override
    public CompletableFuture<Message> complete(List<Message> messages) {
        try {
            ObjectNode body = objectMapper.createObjectNode();
            body.put("model", model);

            ArrayNode messagesNode = body.putArray("messages");
            for (Message msg : messages) {
                ObjectNode msgNode = messagesNode.addObject();
                msgNode.put("role", msg.getRole());
                msgNode.put("content", msg.contentString());
            }

            String requestBody = objectMapper.writeValueAsString(body);
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(baseUrl + "/chat/completions"))
                    .header("Content-Type", "application/json")
                    .header("Authorization", "Bearer " + apiKey)
                    .POST(HttpRequest.BodyPublishers.ofString(requestBody))
                    .timeout(Duration.ofSeconds(60))
                    .build();

            return httpClient.sendAsync(request, HttpResponse.BodyHandlers.ofString())
                    .thenApply(response -> {
                        try {
                            JsonNode root = objectMapper.readTree(response.body());
                            String content = root.at("/choices/0/message/content").asText("");
                            return Message.of("assistant", content);
                        } catch (Exception e) {
                            log.error("failed to parse OpenAI response: {}", e.getMessage());
                            return Message.of("assistant", "Error: failed to parse response");
                        }
                    });
        } catch (Exception e) {
            return CompletableFuture.failedFuture(e);
        }
    }

    @Override
    public String getModel() { return model; }
}
